planner/core/find_best_task.go

```diff
@@ -424,15 +425,37 @@ func (ds *DataSource) findBestTask(prop *property.PhysicalProperty) (t task, err
 
 	t = invalidTask
 	candidates := ds.skylinePruning(prop)
+
+	var taskOfIndexMerge task = invalidTask
+	onlyTblName := ds.indexMergeHints != nil &&
+		(len(ds.indexMergeHints) == 1 && ds.indexMergeHints[0].IndexNames == nil)
+	currentUsedIndexCount := 0
+
 	for _, candidate := range candidates {
 		path := candidate.path
 		if path.partialIndexPaths != nil {
 			idxMergeTask, err := ds.convertToIndexMergeScan(prop, candidate)
 			if err != nil {
 				return nil, err
 			}
-			if idxMergeTask.cost() < t.cost() {
-				t = idxMergeTask
+			// If index merge hint only specifies the table name, we should return the index merge
+			// path with minimal cost. if index merge hint specifies the table name and indexes, we
+			// should return the index merge path with the most covered indexes and minimal cost.
+			if onlyTblName {
+				if idxMergeTask.cost() < taskOfIndexMerge.cost() {
+					taskOfIndexMerge = idxMergeTask
+				}
+				continue
+			}
+			if len(path.partialIndexPaths) > currentUsedIndexCount {
+				taskOfIndexMerge = idxMergeTask
+				currentUsedIndexCount = len(path.partialIndexPaths)
+				continue
+			} else if len(path.partialIndexPaths) == currentUsedIndexCount {
+				if idxMergeTask.cost() < taskOfIndexMerge.cost() {
+					taskOfIndexMerge = idxMergeTask
+				}
+				continue
```

Comment:
A more proper position to impose the index merge hint IMO should be in `DataSource::DeriveStats` or `generateIndexMergeOrPaths`, in which we can prune all other regular access paths if hint is specified and index merge path exists. Also, we can just rely on cost to decide the best index merge path, without comparing the number of indexes covered actually?